### PR TITLE
Feature 6 - dataset generation

### PR DESCRIPTION
**Pull request by Shivesh777**
Contains code to generate all four kinds of datasets.

### CODE REVIEW

File: src/main/java/generateDatasets.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
+import java.lang.Math;
+import java.util.ArrayList;
+import java.util.Random;
+
+
+public class generateDatasets {
+
+    public static void main(String[] args) {
```
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:8]:
```suggestion
    
```

File: src/main/java/generateDatasets.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
```
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:9]:
```suggestion

```

File: src/main/java/generateDatasets.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
```
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:10]:
```suggestion
    
```

File: src/main/java/generateDatasets.java (hunk 1 above)
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:8]: Maybe no need for a main method at all

File: src/main/java/generateDatasets.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            x1.add(xp);
+            y1.add(yp);
+        }
+        dataset1.add(x1);
+        dataset1.add(y1);
+        dataset.add(dataset1);
+        return dataset;
+    }
+
+    public static ArrayList<ArrayList<ArrayList<Double>>> generateClusters() {
```
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:48]: In general, we want `generateNouns` to be the method we call to create the dataset right?

File: src/main/java/generateDatasets.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            y.add(yValue);
+        }
+
+        dataset.add(x);
+        dataset.add(y);
+
+        return dataset;
+    }
+
+    public static ArrayList<ArrayList<ArrayList<Double>>> generateSpiralDatasets() {
```
Comment (Rishit-dagli) [inline src/main/java/generateDatasets.java:126]: So we want our datasets looking like:

$((x,y),\hat{y})$

or maybe like (so we have no ragged tensors):

$((x,y), (\hat{y}))$

where $x$, $y$ are co-ordinates and $\hat{y}$ is our label.

How do our outputs currently look from this?

**Review by Rishit-dagli — CHANGES_REQUESTED** [review]
- One refactor in general, shouldn't take too long would be a separate class for each of these dataset types each with two methods: `generateInstance` and `generate` (this method would be what currently `generateNouns` methods right now). We could thus also have an interface for each of these dataset classes. What do you think about this?

- Also, one minor could we also have `javadocs` for this (it's all in a single repo), maybe one thing we could do is have this as a separate PR too.

- Could you please lint this with https://github.com/google/google-java-format, I'll also make a CI workflow so all commits are checked for this

\cc Shivesh777

**Comment by Rishit-dagli** [comment]
ok lol (I guess I love k8ci-robot way too much), cc: @Shivesh777

File: src/main/java/generateDatasets.java (hunk 5 above)
Comment (Shivesh777) [reply src/main/java/generateDatasets.java:125]: Currently I have, [ { (x), (y) }, { (x), (y) } ].
How do you want it?

File: src/main/java/generateDatasets.java (hunk 4 above)
Comment (Shivesh777) [reply src/main/java/generateDatasets.java:43]: Yes, the singular ones are helper function.

File: src/main/java/generateDatasets.java (hunk 5 above)
Comment (Rishit-dagli) [reply src/main/java/generateDatasets.java:121]: whats the second pair of x and y for?

Comment (Shivesh777) [reply src/main/java/generateDatasets.java:121]: Each pair of coordinates correspond to their respective weight, in this case, 0 and 1.

**Comment by Shivesh777** [comment]
> * One refactor in general, shouldn't take too long would be a separate class for each of these dataset types each with two methods: `generateInstance` and `generate` (this method would be what currently `generateNouns` methods right now). We could thus also have an interface for each of these dataset classes. What do you think about this?
> 
>     * Also, one minor could we also have `javadocs` for this (it's all in a single repo), maybe one thing we could do is have this as a separate PR too.
> 
>     * Could you please lint this with https://github.com/google/google-java-format, I'll also make a CI workflow so all commits are checked for this
> 
> 
> \cc Shivesh777

Yes generateNouns is the correct way to call it. I can make it into separate classes but in our meeting prior to implementation we had decided not to go ahead with different classes.

I will work on the docs in a separate PR.

**Review by Rishit-dagli — APPROVED** [review]
> Yes generateNouns is the correct way to call it. I can make it into separate classes but in our meeting prior to implementation we had decided not to go ahead with different classes.

In that case, lgtm from me

**Review by Ford-Robert — APPROVED** [review]
Looks great to me!

**Pull request merged**